packages/roosterjs-editor-core/lib/undo/Undo.ts

```diff
@@ -0,0 +1,226 @@
+import UndoSnapshots from './UndoSnapshots';
+import containsImage from './containsImage';
+import {
+    ContentChangedEvent,
+    PluginDomEvent,
+    PluginEvent,
+    PluginEventType,
+} from 'roosterjs-editor-types';
+import Editor from '../editor/Editor';
+import UndoService from '../editor/UndoService';
+import { buildSnapshot, restoreSnapshot} from './snapshotUtils';
+
+const KEY_BACKSPACE = 8;
+const KEY_DELETE = 46;
+const KEY_SPACE = 32;
+const KEY_ENTER = 13;
+const KEY_PAGEUP = 33;
+const KEY_DOWN = 40;
+
+/**
+ * Provides snapshot based undo service for Editor
+ */
+export default class Undo implements UndoService {
+    private editor: Editor;
+    private isRestoring: boolean;
+    private hasNewContent: boolean;
+    private undoSnapshots: UndoSnapshots;
+    private lastKeyPress: number;
+
+    /**
+     * Create an instance of Undo
+     * @param preserveSnapshots True to preserve the snapshots after dispose, this allows
+     * this object to be reused when editor is disposed and created again
+     */
+    constructor(
+        private preserveSnapshots?: boolean) {}
+
+    /**
+     * Initialize this plugin. This should only be called from Editor
+     * @param editor Editor instance
+     */
+    public initialize(editor: Editor): void {
+        this.editor = editor;
+
+        // Add an initial snapshot if snapshotsManager isn't created yet
+        if (!this.undoSnapshots) {
+            this.addUndoSnapshot();
+        }
+    }
+
+    /**
+     * Dispose this plugin
+     */
+    public dispose() {
+        this.editor = null;
+
+        if (!this.preserveSnapshots) {
+            this.clear();
+        }
+    }
+
+    /**
+     * Handle events triggered from editor
+     * @param event PluginEvent object
+     */
+    public onPluginEvent(event: PluginEvent): void {
+        // if editor is in IME, don't do anything
+        if (this.editor.isInIME()) {
+            return;
+        }
+
+        switch (event.eventType) {
+            case PluginEventType.KeyDown:
+                this.onKeyDown(event as PluginDomEvent);
+                break;
+            case PluginEventType.KeyPress:
+                this.onKeyPress(event as PluginDomEvent);
+                break;
+            case PluginEventType.CompositionEnd:
+                this.clearRedoForInput();
+                break;
+            case PluginEventType.ContentChanged:
+                if ((<ContentChangedEvent>event).source != 'Undo' && !this.isRestoring) {
```

Comment:
What's the difference between "<ContentChangedEvent>event).source != 'Undo'" and "!this.isRestoring" ?  Doesn't event source is undo meaning it's restoring a snapshot?

Reply:
Removed the check for 'Undo'. And remove the code to trigger ContentChanged event since setContent is already triggering one.